### PR TITLE
Updates test suite to be used with pytest-run-parallel

### PR DESCRIPTION
**Pull request by ngoldbaum**
## Description

Supercedes #7678.

Updates the test suite to be runnable under pytest-run-parallel.

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
The project now has testing on the free-threaded-build using pytest-run-parallel to detect thread safety issues.
```

### CODE REVIEW

File: src/skimage/conftest.py [hunk 1]
```diff
@@ -3,9 +3,25 @@
...
+except Exception:
+    PARALLEL_RUN_AVAILABLE = False
+
+
 import pytest
 
 
+def pytest_configure(config):
+    if not PARALLEL_RUN_AVAILABLE:
+        config.addinivalue_line(
```
Comment (stefanv) [inline src/skimage/conftest.py]: Looks like the marker is defined optionally here. So, why is `pytest_run_parallel` not available on macos?

File: .github/workflows/test-linux.yaml [hunk 2]
```diff
@@ -121,12 +121,19 @@ jobs:
...
         env:
           # A lazy loader configuration parameter
           EAGER_IMPORT: ${{ matrix.EAGER_IMPORT }}
         run: |
           $PYTEST --doctest-plus --showlocals --pyargs skimage ./tests
 
+      - name: Run multithreaded tests (3.14t only)
+        if: ${{ matrix.python-version == '3.14t' }}
+        run: |
+          pip install pytest-run-parallel
```
Comment (stefanv) [inline .github/workflows/test-linux.yaml]: OK, so this would be the reason why it's not available on macos; but why does any other target pass?

Comment (stefanv) [reply .github/workflows/test-linux.yaml]: Looks like marker names are not actually enforced; you can do `pytest.mark.anything_you_like`.

Comment (stefanv) [reply .github/workflows/test-linux.yaml]: I *do* see the warning (which gets turned into error) on a simple reproducer, so I'm still a bit confused why it isn't showing up on every single target.

Comment (stefanv) [reply .github/workflows/test-linux.yaml]: We do have `strict-markers` enabled in our configuration—which *should* be picked up on Windows and Linux too.

Comment (stefanv) [reply .github/workflows/test-linux.yaml]: And now macOS is the only one failing, so looking like it is not picking up the same config.

File: src/skimage/conftest.py (hunk 1 above)
Comment (ngoldbaum) [reply src/skimage/conftest.py]: This is how we set it up in NumPy, FWIW. I didn't want to make `pytest_run_parallel` a required test dependency for everyone who wants to test scikit-image.

Comment (ngoldbaum) [reply src/skimage/conftest.py]: But I think you did identify the issue: scikit-image has two different conftest.py files! I totally missed that.

File: src/skimage/io/collection.py [hunk 3]
```diff
@@ -304,28 +306,31 @@ def __getitem__(self, n):
...
-                    self.data[idx] = self.load_func(self.files[n], **kwargs)
-                self._cached = n
-
-            return self.data[idx]
+                        # Account for functions that do not accept an img_num kwarg
+                        except TypeError as e:
+                            if "unexpected keyword argument 'img_num'" in str(e):
+                                del kwargs['img_num']
+                                self.data[idx] = self.load_func(fname, **kwargs)
+                            else:
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply src/skimage/io/collection.py:327]: This is commenting on pre-existing issues, so I'm ignoring it.

File: src/skimage/segmentation/morphsnakes.py [hunk 4]
```diff
@@ -78,9 +79,7 @@ def inf_sup(u):
     return np.stack(dilations, axis=0).min(0)
 
 
-_curvop = _fcycle(
-    [lambda u: sup_inf(inf_sup(u)), lambda u: inf_sup(sup_inf(u))]  # SIoIS
-)  # ISoSI
+_curvop = threading.local()
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply src/skimage/segmentation/morphsnakes.py:82]: The suggested diff is nonsense. While it's correct that making this state be thread-local doesn't fix the fact that API calls won't be independent, that's also the pre-existing single-threaded behavior.

A better way to fix this is probably to store the `_fcycle` state in a context variable and use `with` statements in the implementation to manage the context state. One way to think of context variables is that they're "souped up" thread locals that are async-local as well as thread-local and that work syntactically with the Python `with` statement.

I could do that refactor, but I'd rather not because the preexisting behavior is probably fine for most uses.

File: tests/skimage/io/test_plugin.py [hunk 5]
```diff
@@ -33,6 +33,7 @@ def test_failed_use():
         manage_plugins.use_plugin('asd')
 
 
+@pytest.mark.thread_unsafe("Tests monkeypatch global plugins state")
 @pytest.mark.skipif(not has_mpl, reason="matplotlib not installed")
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply tests/skimage/io/test_plugin.py]: I disagree that a module-level mark is better, since that means the other tests in this file which are currently being run won't be run.

File: tests/skimage/segmentation/test_random_walker.py [hunk 6]
```diff
@@ -8,33 +8,35 @@
...
+    data = np.zeros((lx, ly)) + 0.1 * rng.randn(lx, ly)
     small_l = int(lx // 5)
     data[
         lx // 2 - small_l : lx // 2 + small_l, ly // 2 - small_l : ly // 2 + small_l
     ] = 1
     data[
         lx // 2 - small_l + 1 : lx // 2 + small_l - 1,
         ly // 2 - small_l + 1 : ly // 2 + small_l - 1,
-    ] = 0.1 * np.random.randn(2 * small_l - 2, 2 * small_l - 2)
+    ] = 0.1 * rng.randn(2 * small_l - 2, 2 * small_l - 2)
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply tests/skimage/segmentation/test_random_walker.py:25]: @stefanv or another reviewer, please let me know if you want me to do this

File: tests/skimage/conftest.py [hunk 7]
```diff
@@ -5,3 +5,19 @@
...
+except Exception:
+    PARALLEL_RUN_AVAILABLE = False
+
+
+def pytest_configure(config):
+    if not PARALLEL_RUN_AVAILABLE:
+        config.addinivalue_line(
+            "markers",
+            "thread_unsafe: mark the test function as single-threaded",
+        )
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply tests/skimage/conftest.py]: This is a real issue.

File: .github/workflows/test-linux.yaml (hunk 2 above)
Comment (ngoldbaum) [reply .github/workflows/test-linux.yaml]: I think this came down to different `conftest.py` processing orders on different architectures, or something along those lines. Moving the configuration to `test/conftest.py` seems to fix everything.

File: tests/skimage/metrics/test_structural_similarity.py [hunk 8]
```diff
@@ -6,28 +6,28 @@
 from skimage._shared.utils import _supported_float_type
 from skimage.metrics import structural_similarity
 
-np.random.seed(5)
 cam = data.camera()
 sigma = 20.0
-cam_noisy = np.clip(cam + sigma * np.random.randn(*cam.shape), 0, 255)
+rng = np.random.RandomState(5)
```
Comment (stefanv) [inline tests/skimage/metrics/test_structural_similarity.py]: Why not `np.random.default_rng(5)`?

**Comment by stefanv** [comment]
A good idea to use local RNG states, as that would also presumably help pave the way towards a parallelized test suite, using pytest-xdist or similar.

File: tests/skimage/metrics/test_structural_similarity.py (hunk 8 above)
Comment (ngoldbaum) [reply tests/skimage/metrics/test_structural_similarity.py]: Becuase `default_rng` isn't subject to the strict backward compatibility guarantees from RandomState. See https://github.com/numpy/numpy/pull/29729 for where we did a similar refactor in NumPy. Robert Kern wrote the new RNG infrastructure so I generally defer to his advice on this stuff.

Comment (stefanv) [reply tests/skimage/metrics/test_structural_similarity.py]: In an ideal world, we wouldn't even set any seeds in our test suite; but we do it to avoid running into edge cases. I think any seed that works is fine (and if you find one that isn't, then we want to know about that too!).

File: src/_skimage2/_shared/testing.py [hunk 9]
```diff
@@ -212,10 +216,11 @@ def fetch(data_filename, prefix=None):
         Path of the local file, possibly pointing to a remote location.
 
     """
-    try:
-        return _fetch(data_filename, prefix=prefix)
-    except (ConnectionError, ModuleNotFoundError):
-        pytest.skip(f'Unable to download {data_filename}', allow_module_level=True)
+    with _FETCH_LOCK:
```
Comment (stefanv) [inline src/_skimage2/_shared/testing.py]: Why can't we download in multiple threads?

File: tests/skimage/feature/test_util.py [hunk 10]
```diff
@@ -151,18 +151,17 @@ def test_plot_matched_features(shapes):
         matches=matches,
         alignment='vertical',
     )
-    plt.close()
+    buf = io.BytesIO()
```
Comment (stefanv) [inline tests/skimage/feature/test_util.py:154]: Is this used anywhere?

File: tests/skimage/io/test_imageio.py [hunk 11]
```diff
@@ -65,7 +66,12 @@ def test_imsave_roundtrip(self, shape, dtype, tmp_path):
             min_, max_, endpoint=True, num=np.prod(shape), dtype=dtype
         )
         expected = expected.reshape(shape)
-        file_path = tmp_path / "roundtrip.png"
+        if hasattr(threading, 'get_native_id'):
+            tmp_dir = tmp_path / str(threading.get_native_id())
```
Comment (stefanv) [inline tests/skimage/io/test_imageio.py]: Out of curiosity, what is this for?

File: tests/skimage/segmentation/test_random_walker.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefanv) [reply tests/skimage/segmentation/test_random_walker.py:25]: No, I think fine as is.

**Review by stefanv — APPROVED** [review]

**Comment by stefanv** [comment]
@greptile review please

**Review by stefanv — APPROVED** [review]
It would be helpful to have a sentence somewhere for the maintainers, so that if they run into this pytest-run-parallel thing, they know what it is about. Otherwise looks great, thanks!

**Comment by ngoldbaum** [comment]
> It would be helpful to have a sentence somewhere for the maintainers, so that if they run into this pytest-run-parallel thing, they know what it is about. Otherwise looks great, thanks!

Thanks! I'll do that and fix the conflicts here once the RNG PR is merged.

**Comment by ngoldbaum** [comment]
I finally came back to this. Now that the other PRs are merged the diff is much smaller here. I'm going to stabilize CI then add developer docs.

File: tests/skimage/feature/test_util.py (hunk 10 above)
Comment (ngoldbaum) [reply tests/skimage/feature/test_util.py:154]: in the next line :)

File: src/_skimage2/_shared/testing.py (hunk 9 above)
Comment (ngoldbaum) [reply src/_skimage2/_shared/testing.py]: Good call, seems not to be necessary in my local testing.

File: tests/skimage/io/test_imageio.py (hunk 11 above)
Comment (ngoldbaum) [reply tests/skimage/io/test_imageio.py]: It's to work around a thread-safety issue from an old version of pytest-run-parallel. Thank for pointing this out, it can be deleted now.

**Comment by ngoldbaum** [comment]
@stefanv I think this is ready for another code review pass.

That said, I saw one Linux failure I can't easily explain: https://github.com/scikit-image/scikit-image/actions/runs/25746117592/job/75610100370

Looking at the implementation of `warp`, my best guess is that this is an OpenBLAS issue maybe? I can't reproduce it on my Mac.

File: tests/skimage/util/test_dtype.py [hunk 12]
```diff
@@ -191,6 +192,7 @@ def test_float_conversion_dtype():
         assert y.dtype == np.dtype(dtype_out)
 
 
+@pytest.mark.thread_unsafe
```
Comment (stefanv) [inline tests/skimage/util/test_dtype.py:195]: I like the pattern you use below where you explain why; if you feel up to it everywhere.

File: CONTRIBUTING.rst [hunk 13]
```diff
@@ -365,6 +370,77 @@ detailing the test coverage::
   ...
 
 
+Multithreaded Testing
+---------------------
+
+Scikit-image supports the free-threaded build and we endeavor to ensure the
```
Comment (stefanv) [inline CONTRIBUTING.rst]:
```suggestion
scikit-image supports the free-threaded build and we endeavor to ensure the
```

File: CONTRIBUTING.rst [hunk 14]
```diff
@@ -365,6 +370,77 @@ detailing the test coverage::
...
+
+Another reason to mark tests as thread-unsafe is because a test spawns a thread
+or process pool. Usually, tests should only use only one level of parallelism
+to avoid CPU oversubscription.
+
+Note that ``pytest-run-parallel`` marks many standard
+library functions and built-in pytest fixtures as thread-unsafe
+automatically. See the ``pytest-run-parallel`` `README
+<https://github.com/Quansight-Labs/pytest-run-parallel#caveats>`_
+for more information.
```
Comment (stefanv) [inline CONTRIBUTING.rst:424]: These notes are great, thanks.

**Review by stefanv — APPROVED** [review]
Very nice, thanks Nathan. I hope that CI target doesn't take too long running "many instances" of the tests in parallel :)

I need to look at the failure; never seen that one around before 🤔

File: tests/skimage/feature/test_util.py (hunk 10 above)
Comment (stefanv) [reply tests/skimage/feature/test_util.py:154]: LOL what happened here

**Comment by stefanv** [comment]
Thanks, @ngoldbaum!
